[c10d][fr] flight recorder improvements

c-p-i-o opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #143446

Summary:
1. Flight recorder dumps are now automatically dumped by default upon
   timeout or exception. Users don't need to opt-in.
2. Change default dump location to running user's home directory
   `.cache` folder.

Test Plan:
1. Tested locally by running the crash program from flight recorder
   tutorial page.
   https://pytorch.org/tutorials/prototype/flight_recorder_tutorial.html#an-end-to-end-example
2. Noted that flight recorder files were correctly created.
❯ pwd
/home/cpio/.cache/fr_trace
❯ ls
nccl_trace_rank_0  nccl_trace_rank_1

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

Differential Revision: [D67363720](https://our.internmc.facebook.com/intern/diff/D67363720)

c-p-i-o committed:
[c10d][fr] flight recorder improvements

Summary:
1. Flight recorder dumps are now automatically dumped by default upon
   timeout or exception. Users don't need to opt-in.
2. Change default dump location to running user's home directory
   `.cache` folder.

Test Plan:
1. Tested locally by running the crash program from flight recorder
   tutorial page.
   https://pytorch.org/tutorials/prototype/flight_recorder_tutorial.html#an-end-to-end-example
2. Noted that flight recorder files were correctly created.
❯ pwd
/home/cpio/.cache/fr_trace
❯ ls
nccl_trace_rank_0  nccl_trace_rank_1

[ghstack-poisoned]

c-p-i-o committed:
Update on "[c10d][fr] flight recorder improvements"


Summary:
1. Flight recorder dumps are now automatically dumped by default upon
   timeout or exception. Users don't need to opt-in.
2. Change default dump location to running user's home directory
   `.cache` folder.

Test Plan:
1. Tested locally by running the crash program from flight recorder
   tutorial page.
   https://pytorch.org/tutorials/prototype/flight_recorder_tutorial.html#an-end-to-end-example
2. Noted that flight recorder files were correctly created.
❯ pwd
/home/cpio/.cache/fr_trace
❯ ls
nccl_trace_rank_0  nccl_trace_rank_1

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

c-p-i-o commented:
@c-p-i-o has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67363720).

c-p-i-o committed:
Update on "[c10d][fr] flight recorder improvements"


Summary:
1. Flight recorder dumps are now automatically dumped by default upon
   timeout or exception. Users don't need to opt-in.
2. Change default dump location to running user's home directory
   `.cache` folder.

Test Plan:
1. Tested locally by running the crash program from flight recorder
   tutorial page.
   https://pytorch.org/tutorials/prototype/flight_recorder_tutorial.html#an-end-to-end-example
2. Noted that flight recorder files were correctly created.
❯ pwd
/home/cpio/.cache/fr_trace
❯ ls
nccl_trace_rank_0  nccl_trace_rank_1

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k

Differential Revision: [D67363720](https://our.internmc.facebook.com/intern/diff/D67363720)

[ghstack-poisoned]

c-p-i-o commented:
@c-p-i-o has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67363720).

c-p-i-o committed:
Update on "[c10d][fr] flight recorder improvements"


Summary:
1. Flight recorder dumps are now automatically dumped by default upon
   timeout or exception. Users don't need to opt-in.
2. Change default dump location to running user's home directory
   `.cache` folder.

Test Plan:
1. Tested locally by running the crash program from flight recorder
   tutorial page.
   https://pytorch.org/tutorials/prototype/flight_recorder_tutorial.html#an-end-to-end-example
2. Noted that flight recorder files were correctly created.
❯ pwd
/home/cpio/.cache/fr_trace
❯ ls
nccl_trace_rank_0  nccl_trace_rank_1

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k

Differential Revision: [D67363720](https://our.internmc.facebook.com/intern/diff/D67363720)

[ghstack-poisoned]

c-p-i-o commented:
@c-p-i-o has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67363720).

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -58,12 +58,6 @@ static std::vector<std::string> TORCH_NCCL_ASYNC_ERROR_HANDLING = {
-// timeout is enabled. This variable must be set together with
-// TORCH_NCCL_ENABLE_MONITORING=1 and TORCH_NCCL_TRACE_BUFFER_SIZE > 0.
-static std::vector<std::string> TORCH_NCCL_DUMP_ON_TIMEOUT = {
-    "TORCH_NCCL_DUMP_ON_TIMEOUT"};
```
hm, change the default to 1 but don't delete the env in case someone wants to disable? we can clean up the env later.

or are you changing so that one of the other ENVs is now the way to disable? 
 - it's not OK to make the only way to disable dumping to disk be to set TRACE_BUFFER_SIZE=0, since some users might only use in-memory + http server

c-p-i-o replied:
Ack. I put back the flag.
We'll do a cleanup pass later.

wconstab commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -4189,7 +4188,6 @@ def setUp(self):
             "TORCH_NCCL_ENABLE_TIMING"
         ] = "0"  # see 'timing_enabled' parametrized tests
         os.environ["TORCH_NCCL_TRACE_BUFFER_SIZE"] = "1000"
-        os.environ["TORCH_NCCL_DUMP_ON_TIMEOUT"] = "1"
```
it's a bit concerning that we'd have to update any tests at all.

(except for one case: if the test used to rely on /tmp/ as the write location without setting the env itself, then the test had a bug and needs to be updated to set the write location)

otherwise, the change in the PR should not affect existing tests should it?

c-p-i-o replied:
This was me removing that flag. I've put back the flag - so existing tests are now unaffected.

wconstab commented on torch/csrc/distributed/c10d/FlightRecorder.cpp:
```diff
@@ -108,6 +113,45 @@ control_plane::RegisterHandler jsonDumpHandler{
+#ifdef _WIN32
+  int ret = _mkdir(dir.c_str());
+#else
+  int ret = mkdir(dir.c_str(), S_IRWXU | S_IRWXG | S_IRWXO);
```
UGO? same as 777? (should the default be just U or UG?)

c-p-i-o committed:
Update on "[c10d][fr] flight recorder improvements"


Summary:
1. Flight recorder dumps are now automatically dumped by default upon
   timeout or exception. Users don't need to opt-in.
2. Change default dump location to running user's home directory
   `.cache` folder.

Test Plan:
1. Tested locally by running the crash program from flight recorder
   tutorial page.
   https://pytorch.org/tutorials/prototype/flight_recorder_tutorial.html#an-end-to-end-example
2. Noted that flight recorder files were correctly created.
❯ pwd
/home/cpio/.cache/fr_trace
❯ ls
nccl_trace_rank_0  nccl_trace_rank_1

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k

Differential Revision: [D67363720](https://our.internmc.facebook.com/intern/diff/D67363720)

[ghstack-poisoned]

c-p-i-o commented:
@c-p-i-o has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67363720).

d4l3k reviewed (approved):
generally seems reasonable -- should we add a unit test for this?

c-p-i-o commented on torch/csrc/distributed/c10d/FlightRecorder.cpp:
```diff
@@ -131,8 +175,16 @@ void DebugInfoWriter::write(const std::string& trace) {
+    // Attempt to write to running user's HOME directory cache folder - if it
+    // exists.
+    auto homeDir = getCvarString({"HOME"}, "/tmp");
+    std::string cacheDirPath = homeDir + "/.cache/fr_trace";
```
Will change to `torch`.

c-p-i-o committed:
Update on "[c10d][fr] flight recorder improvements"


Summary:
1. Flight recorder dumps are now automatically dumped by default upon
   timeout or exception. Users don't need to opt-in.
2. Change default dump location to running user's home directory
   `.cache` folder.

Test Plan:
1. Tested locally by running the crash program from flight recorder
   tutorial page.
   https://pytorch.org/tutorials/prototype/flight_recorder_tutorial.html#an-end-to-end-example
2. Noted that flight recorder files were correctly created.
❯ pwd
/home/cpio/.cache/fr_trace
❯ ls
nccl_trace_rank_0  nccl_trace_rank_1

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k

Differential Revision: [D67363720](https://our.internmc.facebook.com/intern/diff/D67363720)

[ghstack-poisoned]

c-p-i-o commented:
@c-p-i-o has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67363720).

c-p-i-o committed:
Update on "[c10d][fr] flight recorder improvements"


Summary:
1. Flight recorder dumps are now automatically dumped by default upon
   timeout or exception. Users don't need to opt-in.
2. Change default dump location to running user's home directory
   `.cache` folder.

Test Plan:
1. Tested locally by running the crash program from flight recorder
   tutorial page.
   https://pytorch.org/tutorials/prototype/flight_recorder_tutorial.html#an-end-to-end-example
2. Noted that flight recorder files were correctly created.
❯ pwd
/home/cpio/.cache/fr_trace
❯ ls
nccl_trace_rank_0  nccl_trace_rank_1

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k

Differential Revision: [D67363720](https://our.internmc.facebook.com/intern/diff/D67363720)

[ghstack-poisoned]

c-p-i-o commented:
@c-p-i-o has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67363720).

c-p-i-o commented:
@c-p-i-o has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67363720).

c-p-i-o committed:
Update on "[c10d][fr] flight recorder improvements"


Summary:
1. Flight recorder dumps are now automatically dumped by default upon
   timeout or exception. Users don't need to opt-in.
2. Change default dump location to running user's home directory
   `.cache` folder.

Test Plan:
1. Tested locally by running the crash program from flight recorder
   tutorial page.
   https://pytorch.org/tutorials/prototype/flight_recorder_tutorial.html#an-end-to-end-example
2. Noted that flight recorder files were correctly created.
❯ pwd
/home/cpio/.cache/fr_trace
❯ ls
nccl_trace_rank_0  nccl_trace_rank_1

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k

Differential Revision: [D67363720](https://our.internmc.facebook.com/intern/diff/D67363720)

[ghstack-poisoned]

c-p-i-o commented:
@c-p-i-o has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67363720).

c-p-i-o committed:
Update on "[c10d][fr] flight recorder improvements"


Summary:
1. Flight recorder dumps are now automatically dumped by default upon
   timeout or exception. Users don't need to opt-in.
2. Change default dump location to running user's home directory
   `.cache` folder.

Test Plan:
1. Tested locally by running the crash program from flight recorder
   tutorial page.
   https://pytorch.org/tutorials/prototype/flight_recorder_tutorial.html#an-end-to-end-example
2. Noted that flight recorder files were correctly created.
❯ pwd
/home/cpio/.cache/fr_trace
❯ ls
nccl_trace_rank_0  nccl_trace_rank_1

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k

Differential Revision: [D67363720](https://our.internmc.facebook.com/intern/diff/D67363720)

[ghstack-poisoned]

c-p-i-o commented:
@c-p-i-o has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67363720).

c-p-i-o commented:
@pytorchbot merge
